File: src/lib/accounts.test.ts
import { describe, expect, it } from 'vitest';
import { parseEntries, parseLine, type ParsedEntry } from './accounts';
import { generateTotp } from './totp';

const asAccount = (entry: ParsedEntry) => {
  if (entry.kind !== 'account') {
    throw new Error(`Erwartet wurde ein Konto, bekommen: ${entry.message}`);
  }
  return entry.account;
};

describe('parseLine — rohe Base32-Secrets', () => {
  it('nimmt ein nacktes Secret an', () => {
    const account = asAccount(parseLine('JBSWY3DPEHPK3PXP'));
    expect(account.secret).toHaveLength(10);
    expect(account.issuer).toBeUndefined();
    expect(account.algorithm).toBe('SHA-1');
    expect(account.digits).toBe(6);
    expect(account.period).toBe(30);
  });

  it('nimmt ein Secret mit Leerzeichen und Kleinbuchstaben an', () => {
    expect(asAccount(parseLine('jbsw y3dp ehpk 3pxp')).secret).toEqual(
      asAccount(parseLine('JBSWY3DPEHPK3PXP')).secret,
    );
  });

  it('erkennt "Name: SECRET"', () => {
    const account = asAccount(parseLine('GitHub: JBSWY3DPEHPK3PXP'));
    expect(account.issuer).toBe('GitHub');
    expect(account.secret).toEqual(asAccount(parseLine('JBSWY3DPEHPK3PXP')).secret);
  });

  it('trennt am letzten Doppelpunkt, damit der Name selbst einen enthalten darf', () => {
    const account = asAccount(parseLine('Arbeit: GitHub: JBSWY3DPEHPK3PXP'));
    expect(account.issuer).toBe('Arbeit: GitHub');
  });
});

describe('parseLine — otpauth-URIs', () => {
  it('übernimmt Label und Parameter aus der URI', () => {
    const account = asAccount(
      parseLine('otpauth://totp/GitHub:kevin?secret=JBSWY3DPEHPK3PXP&issuer=GitHub&digits=8'),
    );
    expect(account.issuer).toBe('GitHub');
    expect(account.accountName).toBe('kevin');
    expect(account.digits).toBe(8);
  });

  it('decodiert das Secret aus der URI', async () => {
    const fromUri = asAccount(
      parseLine('otpauth://totp/RFC?secret=GEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQ'),
    );
    expect(await generateTotp({ secret: fromUri.secret, unixSeconds: 59 })).toBe('287082');
  });
});

describe('parseLine — Fehler werden zu Ergebnissen, nicht zu Ausnahmen', () => {
  it('meldet ungültige Base32-Zeichen', () => {
    const entry = parseLine('JBSW0Y3DPEHPK3PX');
    expect(entry.kind).toBe('error');
    if (entry.kind === 'error') {
      expect(entry.message).toMatch(/»0«/);
      expect(entry.source).toBe('JBSW0Y3DPEHPK3PX');
    }
  });

  it('meldet kaputte URIs', () => {
    const entry = parseLine('otpauth://totp/Test');
    expect(entry.kind).toBe('error');
    if (entry.kind === 'error') {
      expect(entry.message).toMatch(/secret/);
    }
  });

  it('wirft niemals', () => {
    for (const line of ['', '?', '::::', 'otpauth://', 'otpauth://totp/%', 'A'.repeat(999)]) {
      expect(() => parseLine(line)).not.toThrow();
    }
  });
});

describe('parseEntries — mehrere Zeilen gemischt', () => {
  const input = [
    '# Meine Konten',
    'JBSWY3DPEHPK3PXP',
    '',
    'GitHub: GEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQ',
    'otpauth://totp/ACME:kevin%40example.com?secret=JBSWY3DPEHPK3PXP&issuer=ACME&period=60',
    '   ',
    'kaputt!!!',
  ].join('\n');

  it('überspringt Leerzeilen und Kommentare', () => {
    expect(parseEntries(input)).toHaveLength(4);
  });

  it('verarbeitet rohe Secrets und URIs gemischt', () => {
    const entries = parseEntries(input);
    expect(entries.map((entry) => entry.kind)).toEqual(['account', 'account', 'account', 'error']);
    expect(asAccount(entries[1]!).issuer).toBe('GitHub');
    expect(asAccount(entries[2]!).period).toBe(60);
  });

  it('vergibt für jede Zeile einen eigenen Schlüssel — auch bei Duplikaten', () => {
    const entries = parseEntries('JBSWY3DPEHPK3PXP\nJBSWY3DPEHPK3PXP');
    expect(entries).toHaveLength(2);
    expect(entries[0]!.key).not.toBe(entries[1]!.key);
  });

  it('lässt eine kaputte Zeile die anderen nicht beschädigen', () => {
    const entries = parseEntries('kaputt!!!\nJBSWY3DPEHPK3PXP');
    expect(entries[0]!.kind).toBe('error');
    expect(entries[1]!.kind).toBe('account');
  });

  it('liefert für leeren Text eine leere Liste', () => {
    expect(parseEntries('')).toEqual([]);
    expect(parseEntries('\n\n  \n# nur ein Kommentar\n')).toEqual([]);
  });
});

File: android-native/app/src/test/kotlin/io/github/keco216/clockwork/core/AccountsTest.kt
package io.github.keco216.clockwork.core

import org.junit.Assert.assertArrayEquals
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Portiert aus `src/lib/accounts.test.ts`.
 */
class AccountsTest {

    private fun account(entry: ParsedEntry): Account {
        assertTrue("Erwartet wurde ein Konto, bekommen: $entry", entry is ParsedEntry.Ok)
        return (entry as ParsedEntry.Ok).account
    }

    @Test
    fun `nimmt ein nacktes Secret an`() {
        val account = account(parseLine("JBSWY3DPEHPK3PXP"))
        assertEquals(10, account.secret.size)
        assertNull(account.issuer)
        assertEquals(HashAlgorithm.SHA1, account.algorithm)
        assertEquals(6, account.digits)
        assertEquals(30, account.period)
    }

    @Test
    fun `nimmt ein Secret mit Leerzeichen und Kleinbuchstaben an`() {
        assertArrayEquals(
            account(parseLine("JBSWY3DPEHPK3PXP")).secret,
            account(parseLine("jbsw y3dp ehpk 3pxp")).secret,
        )
    }

    @Test
    fun `erkennt Name doppelpunkt SECRET`() {
        val account = account(parseLine("GitHub: JBSWY3DPEHPK3PXP"))
        assertEquals("GitHub", account.issuer)
        assertArrayEquals(account(parseLine("JBSWY3DPEHPK3PXP")).secret, account.secret)
    }

    @Test
    fun `trennt am letzten Doppelpunkt`() {
        // Damit der Name selbst einen enthalten darf — das Secret nie.
        assertEquals("Arbeit: GitHub", account(parseLine("Arbeit: GitHub: JBSWY3DPEHPK3PXP")).issuer)
    }

    @Test
    fun `uebernimmt Label und Parameter aus der URI`() {
        val account = account(
            parseLine("otpauth://totp/GitHub:kevin?secret=JBSWY3DPEHPK3PXP&issuer=GitHub&digits=8"),
        )
        assertEquals("GitHub", account.issuer)
        assertEquals("kevin", account.accountName)
        assertEquals(8, account.digits)
    }

    @Test
    fun `decodiert das Secret aus der URI`() {
        val account = account(parseLine("otpauth://totp/RFC?secret=GEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQ"))
        assertEquals("287082", generateTotp(account.secret, 59.0))
    }

    @Test
    fun `meldet ungueltige Base32-Zeichen als Ergebnis statt als Ausnahme`() {
        val entry = parseLine("JBSW0Y3DPEHPK3PX")
        assertTrue(entry is ParsedEntry.Failed)
        entry as ParsedEntry.Failed
        assertEquals("err.base32.badChar", entry.messageKey)
        assertEquals("0", entry.messageArgs["char"])
        assertEquals("JBSW0Y3DPEHPK3PX", entry.source)
    }

    @Test
    fun `meldet kaputte URIs`() {
        val entry = parseLine("otpauth://totp/Test")
        assertTrue(entry is ParsedEntry.Failed)
        assertEquals("err.uri.noSecret", (entry as ParsedEntry.Failed).messageKey)
    }

    @Test
    fun `parseLine wirft niemals`() {
        for (line in listOf("", "?", "::::", "otpauth://", "otpauth://totp/%", "A".repeat(999))) {
            // Kein assertThrows, sondern der Aufruf selbst: Ein geworfener
            // Fehler liesse den Test scheitern, und genau das ist die Zusage.
            val entry = parseLine(line)
            assertTrue("Zeile \"$line\"", entry is ParsedEntry.Failed || entry is ParsedEntry.Ok)
        }
    }

    private val mixed = listOf(
        "# Meine Konten",
        "JBSWY3DPEHPK3PXP",
        "",
        "GitHub: GEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQ",
        "otpauth://totp/ACME:kevin%40example.com?secret=JBSWY3DPEHPK3PXP&issuer=ACME&period=60",
        "   ",
        "kaputt!!!",
    ).joinToString("\n")

    @Test
    fun `ueberspringt Leerzeilen und Kommentare`() {
        assertEquals(4, parseEntries(mixed).size)
    }

    @Test
    fun `verarbeitet rohe Secrets und URIs gemischt`() {
        val entries = parseEntries(mixed)
        assertEquals(
            listOf(true, true, true, false),
            entries.map { it is ParsedEntry.Ok },
        )
        assertEquals("GitHub", account(entries[1]).issuer)
        assertEquals(60, account(entries[2]).period)
    }

    @Test
    fun `vergibt fuer jede Zeile einen eigenen Schluessel auch bei Duplikaten`() {
        val entries = parseEntries("JBSWY3DPEHPK3PXP\nJBSWY3DPEHPK3PXP")
        assertEquals(2, entries.size)
        assertNotEquals(entries[0].key, entries[1].key)
    }

    @Test
    fun `laesst eine kaputte Zeile die anderen nicht beschaedigen`() {
        val entries = parseEntries("kaputt!!!\nJBSWY3DPEHPK3PXP")
        assertTrue(entries[0] is ParsedEntry.Failed)
        assertTrue(entries[1] is ParsedEntry.Ok)
    }

    @Test
    fun `liefert fuer leeren Text eine leere Liste`() {
        assertEquals(emptyList<ParsedEntry>(), parseEntries(""))
        assertEquals(emptyList<ParsedEntry>(), parseEntries("\n\n  \n# nur ein Kommentar\n"))
    }
}
